rust/worker/src/index/fulltext/types.rs

```diff
@@ -116,227 +50,605 @@ impl FullTextIndex for BlockfileFullTextIndex {
+            .await?;
+        self.frequencies_blockfile_writer
+            .commit::<u32, &str>()?
+            .flush::<u32, u32>()
+            .await?;
+        Ok(())
+    }
+}
+
+pub(crate) struct FullTextIndexReader<'me> {
+    posting_lists_blockfile_reader: BlockfileReader<'me, u32, Int32Array>,
+    frequencies_blockfile_reader: BlockfileReader<'me, u32, u32>,
+    tokenizer: Box<dyn ChromaTokenizer>,
+}
+
+impl<'me> FullTextIndexReader<'me> {
+    fn new(
+        posting_lists_blockfile_reader: BlockfileReader<'me, u32, Int32Array>,
+        frequencies_blockfile_reader: BlockfileReader<'me, u32, u32>,
+        tokenizer: Box<dyn ChromaTokenizer>,
+    ) -> Self {
+        FullTextIndexReader {
+            posting_lists_blockfile_reader,
+            frequencies_blockfile_reader,
+            tokenizer,
+        }
+    }
+
+    async fn search(&mut self, query: &str) -> Result<Vec<i32>, Box<dyn ChromaError>> {
         let binding = self.tokenizer.encode(query);
         let tokens = binding.get_tokens();
 
         // Get query tokens sorted by frequency.
-        let mut token_frequencies = vec![];
+        let mut token_frequencies: Vec<(String, u32)> = vec![];
         for token in tokens {
-            let blockfilekey =
-                BlockfileKey::new("".to_string(), Key::String(token.text.to_string()));
-            let value = self.frequencies_blockfile.get(blockfilekey);
-            match value {
-                Ok(Value::IntValue(frequency)) => {
-                    token_frequencies.push((token.text.to_string(), frequency));
-                }
-                Ok(_) => {
-                    return Ok(vec![]);
-                }
-                Err(_) => {
-                    // TODO error handling from blockfile
-                    return Ok(vec![]);
-                }
+            // TODO better error matching (NotFoundError should return Ok(vec![])) but some others should error.
+            let res = self
+                .frequencies_blockfile_reader
+                .get_by_prefix(token.text.as_str())?;
+            if res.len() == 0 {
+                return Ok(vec![]);
             }
+            if res.len() > 1 {
+                panic!("Multiple tokens found in frequencies blockfile");
+            }
+            let res = res[0];
+            // Throw away the "value" since we store frequencies in the keys.
+            token_frequencies.push((token.text.to_string(), res.1));
         }
-        token_frequencies.sort_by(|a, b| a.1.cmp(&b.1));
+        // TODO sort by frequency. This adds an additional layer of complexity
+        // with repeat characters where we need to keep track of which positions
+        // for the character have been seen/used in the matching algorithm. By
+        // leaving them ordered per the query, we can stick to the more straightforward
+        // but less efficient matching algorithm.
+        // token_frequencies.sort_by(|a, b| a.1.cmp(&b.1));
```

Comment:
Should we do the work to sort by token frequency now? It'll take some time to get it right and test. This bug was lurking in the old implementation (yet another argument for property testing).